libmscore/rendermidi.cpp

```diff
@@ -390,20 +397,105 @@ static void aeolusSetStop(int tick, int channel, int i, int k, bool val, EventMa
       }
 
 //---------------------------------------------------------
-//   collectMeasureEvents
+//   collectProgramChanges
 //---------------------------------------------------------
 
-static void collectMeasureEvents(EventMap* events, Measure* m, Staff* staff, int tickOffset)
+static void collectProgramChanges(EventMap* events, Measure* m, Staff* staff, int tickOffset)
       {
       int firstStaffIdx = staff->idx();
       int nextStaffIdx  = firstStaffIdx + 1;
 
+      //
+      // collect program changes and controller
+      //
+      for (Segment* s = m->first(SegmentType::ChordRest); s; s = s->next(SegmentType::ChordRest)) {
+            for (Element* e : s->annotations()) {
+                  if (!e->isStaffTextBase() || e->staffIdx() < firstStaffIdx || e->staffIdx() >= nextStaffIdx)
+                        continue;
+                  const StaffTextBase* st1 = toStaffTextBase(e);
+                  int tick = s->tick() + tickOffset;
+
+                  Instrument* instr = e->part()->instrument(tick);
+                  for (const ChannelActions& ca : *st1->channelActions()) {
+                        int channel = instr->channel().at(ca.channel)->channel();
+                        for (const QString& ma : ca.midiActionNames) {
+                              NamedEventList* nel = instr->midiAction(ma, ca.channel);
+                              if (!nel)
+                                    continue;
+                              for (MidiCoreEvent event : nel->events) {
+                                    event.setChannel(channel);
+                                    NPlayEvent e1(event);
+                                    e1.setOriginatingStaff(firstStaffIdx);
+                                    if (e1.dataA() == CTRL_PROGRAM)
+                                          events->insert(std::pair<int, NPlayEvent>(tick-1, e1));
+                                    else
+                                          events->insert(std::pair<int, NPlayEvent>(tick, e1));
+                                    }
+                              }
+                        }
+                  if (st1->setAeolusStops()) {
+                        Staff* s1 = st1->staff();
+                        int voice   = 0;
+                        int channel = s1->channel(tick, voice);
+
+                        for (int i = 0; i < 4; ++i) {
+                              static int num[4] = { 12, 13, 16, 16 };
```

Comment:
This one might be silly of me, but a list of numbers like this seems to warrant a comment explaining what they are for just Incase of a future issue.

Reply:
Ah, the code in `collectProgramChanges` was a copy and paste from the old `collectMeasureEvents` code, so I could call it from both of the new collectMeasureEvents functions. I have no idea what the `num` array does. My guess is that they're the ids of Aeolus stops? I'm going to leave it for now, since renaming it might make working with it even more confusing if I rename it incorrectly.